### PR TITLE
add ITB fatality model based on a Bayesian approach

### PR DESCRIPTION
**Pull request by dynaryu**
Both the PAGER and ITB Bayesian fatality model can provide the probability of fatality in magnitude bin as the USGS Pager does.

### CODE REVIEW

File: safe/impact_functions/earthquake/itb_bayesian_earthquake_fatality_model/metadata_definitions.py [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+    count_exposure_unit,
+    hazard_category_single_event,
+    unit_mmi
+)
+
+
+class ITBBayesianFatalityMetadata(ITBFatalityMetadata):
+    """Metadata for PAG Fatality Function.
+
+    .. versionadded:: 2.1
```
Comment (timlinux) [inline safe/impact_functions/earthquake/itb_bayesian_earthquake_fatality_model/metadata_definitions.py]: .. versionadded:: 3.3

File: safe/impact_functions/earthquake/itb_bayesian_earthquake_fatality_model/impact_function.py [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+    itb_earthquake_fatality_model.impact_function import ITBFatalityFunction
+from safe.impact_functions.earthquake\
+    .itb_bayesian_earthquake_fatality_model.metadata_definitions import \
+    ITBBayesianFatalityMetadata
+
+
+class ITBBayesianFatalityFunction(ITBFatalityFunction):
+    # noinspection PyUnresolvedReferences
+    """ ITB fatality model based on a Bayesian approach
+
```
Comment (timlinux) [inline safe/impact_functions/earthquake/itb_bayesian_earthquake_fatality_model/impact_function.py]: .. versionadded:: 3.3

File: safe/impact_functions/earthquake/itb_bayesian_earthquake_fatality_model/test/test_itb_bayesian_earthquake_fatality_model.py [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+from safe.impact_functions.earthquake.itb_bayesian_earthquake_fatality_model\
+    .impact_function import ITBBayesianFatalityFunction
+from safe.test.utilities import test_data_path, get_qgis_app, clip_layers
+from safe.storage.core import read_layer
+from safe.storage.safe_layer import SafeLayer
+
+QGIS_APP, CANVAS, IFACE, PARENT = get_qgis_app()
+
+class TestITBBayesianEarthquakeFatalityFunction(unittest.TestCase):
+    """Test for ITB Bayesian Earthquake on Population Impact Function."""
```
Comment (timlinux) [inline safe/impact_functions/earthquake/itb_bayesian_earthquake_fatality_model/test/test_itb_bayesian_earthquake_fatality_model.py]: .. versionadded:: 3.3

File: safe/impact_functions/earthquake/itb_earthquake_fatality_model/impact_function.py [hunk 4]
```diff
@@ -126,20 +126,52 @@ def __init__(self):
         ])
         self.total_fatalities = None
 
-    def fatality_rate(self, mmi):
+    def compute_fatality_rate(self):
         """ITB method to compute fatality rate.
```
Comment (timlinux) [inline safe/impact_functions/earthquake/itb_earthquake_fatality_model/impact_function.py]: .. versionchanged:: 3.3 - renamed method from fatality_rate to compute_fatality_rate

File: safe/impact_functions/earthquake/itb_earthquake_fatality_model/impact_function.py [hunk 5]
```diff
@@ -126,20 +126,52 @@ def __init__(self):
...
-        return numpy.power(10.0, x * mmi - y)
+        fatality_rate = {mmi: 0 if mmi < 4 else numpy.power(10.0, x * mmi - y)
+                         for mmi in mmi_range}
+        return fatality_rate
+
+    @staticmethod
+    def round_to_sum(prob_array):
+        """
+        Round a list of numbers while maintaining the sum.
+
```
Comment (timlinux) [inline safe/impact_functions/earthquake/itb_earthquake_fatality_model/impact_function.py]: ..versionadded:: 3.3

File: safe/impact_functions/earthquake/itb_earthquake_fatality_model/impact_function.py [hunk 6]
```diff
@@ -296,8 +329,19 @@ def run(self):
             number_of_fatalities[mmi] = fatalities
 
         # Total statistics
+        total_fatalities_raw = numpy.nansum(
+            number_of_fatalities.values(), axis=0)
+
+        # Compute probaility of fatality in each magnitude bin
+        if (self.__class__.__name__ == 'PAGFatalityFunction') or (
```
Comment (timlinux) [inline safe/impact_functions/earthquake/itb_earthquake_fatality_model/impact_function.py]: Hmm this is vulnerable to name changes in the class....I guess our Unit tests will catch breakages, but I wonder if there is another approach we can take...

File: safe/impact_functions/earthquake/itb_earthquake_fatality_model/test/test_itb_earthquake_fatality_model.py [hunk 7]
```diff
@@ -35,6 +35,24 @@ def setUp(self):
         registry.clear()
         registry.register(ITBFatalityFunction)
 
+    def test_compute_fatality_rate(self):
+        impact_function = ITBFatalityFunction.instance()
```
Comment (timlinux) [inline safe/impact_functions/earthquake/itb_earthquake_fatality_model/test/test_itb_earthquake_fatality_model.py]: """Test compute fatality rate.

.. versionadded:: 3.3
"""

File: safe/impact_functions/earthquake/pager_earthquake_fatality_model/test/test_pager_earthquake_fatality_model.py [hunk 8]
```diff
@@ -34,6 +34,48 @@ def setUp(self):
         registry.clear()
         registry.register(PAGFatalityFunction)
 
+    def test_compute_fatality_rate(self):
```
Comment (timlinux) [inline safe/impact_functions/earthquake/pager_earthquake_fatality_model/test/test_pager_earthquake_fatality_model.py]: Jut need docstrings and versionadded annotation for these methods...

**Comment by timlinux** [comment]
Thanks for this @dynaryu - is your idea to replace the IF that we use for realtime to use this implementation, or should we still consider this experimental for now?

**Comment by dynaryu** [comment]
@timlinux The PAGER model is the same model that USGS is currenlty using, so there will be no problem in running as long as there is no bug in my code. The main issue is whether we want to show the probability of fatality in each magnitude either realtime or inasafe, which will require some changes to the layout.

**Comment by Charlotte-Morgan** [comment]
@samnawi - can you please look at this PR. Its a new EQ IF from Hyeuk. 

@davidr007 - any comments on using this for realtime

File: safe/impact_functions/earthquake/itb_earthquake_fatality_model/impact_function.py (hunk 6 above)
Comment (dynaryu) [inline safe/impact_functions/earthquake/itb_earthquake_fatality_model/impact_function.py]: I'm quite open to any suggestion.

**Comment by timlinux** [comment]
@dynaryu Just a note that in order for this IF to appear in the DOCK, WIZARD user interfaces, you should also register it with the IF registry. Please see how the other EQ IF's were registered in ``safe/impact_functions/__init__.py`` and add a similar entry for the Bayesian EQ IF. Thanks!

**Comment by timlinux** [comment]
@dynaryu don't worry if you don't have a chance to make the ``__init__.py`` change requested above - I will do it quickly before merging if needed.

**Comment by lucernae** [comment]
Don't mind this comment, just popping in to make sure I subscribed to the PR

**Comment by timlinux** [comment]
Hi 

@lucernae can you please take over getting this merged to develop?

1) Add the needed bits to register it with the IF registry as per my note above
2) As @samnawi and @davidr007 to review as per @Charlotte-Morgan 's note above
3) add a changelog entry in metadata.py
4) merge to develop
5) add an entry to the visual changelog (I will edit the english after)

**Comment by timlinux** [comment]
I'm going to go ahead and merge this and sort out the registry etc items in a subsequent pull.

**Comment by lucernae** [comment]
@timlinux thank's, I've been too focused out with flood realtime

**Pull request merged**